### ISSUE TITLE
DDP doesn't work with retain_graph = True

### ISSUE THREAD

**Issue by pritamdamania87** [pytorch/pytorch#47260]
## 🐛 Bug

DDP doesn't work with retain_graph = True when trying to run backwards twice through the same model.

## To Reproduce

To replicate, change only def demo_basic(rank, world_size) in https://pytorch.org/tutorials/intermediate/ddp_tutorial.html 2 to the following:

```
def demo_basic(rank, world_size):
    print(f"Running basic DDP example on rank {rank}.")
    setup(rank, world_size)

    # create model and move it to GPU with id rank
    model = ToyModel().to(rank)
    ddp_model = DDP(model, device_ids=[rank])

    loss_fn = nn.MSELoss()
    optimizer = optim.SGD(ddp_model.parameters(), lr=1)
    
    optimizer.zero_grad()
    
    outputs = {}
    outputs['0'] = ddp_model(torch.rand(20, 10))
    outputs['1'] = ddp_model(torch.rand(20, 10))
    outputs['2'] = ddp_model(torch.rand(20, 10))

    labels = torch.rand(20, 5).to(rank)

    for i in range(3):
        print(f"before {i}, rank: {rank}, weight: {ddp_model.module.net1.weight[0][0]}")
                 
        if i < 2:
            loss_fn(outputs[str(i)], labels).backward(retain_graph=True)
        else:
            loss_fn(outputs[str(i)], labels).backward()

        print(f"after {i}, rank: {rank}, weight: {ddp_model.module.net1.weight[0][0]}, grad: {ddp_model.module.net1.weight.grad[0][0]}")
                
    optimizer.step()
    print(f"last, rank: {rank}, weight: {ddp_model.module.net1.weight[0][0]}, grad: {ddp_model.module.net1.weight.grad[0][0]}")
    
    cleanup()
```
and the output is:
```
before 0, rank: 0, weight: 0.1450435221195221
before 0, rank: 3, weight: 0.1450435221195221
before 0, rank: 1, weight: 0.1450435221195221
before 0, rank: 2, weight: 0.1450435221195221
after 0, rank: 0, weight: 0.1450435221195221, grad: -0.018003715202212334
before 1, rank: 0, weight: 0.1450435221195221
after 0, rank: 3, weight: 0.1450435221195221, grad: -0.018003715202212334
after 0, rank: 1, weight: 0.1450435221195221, grad: -0.018003715202212334
after 0, rank: 2, weight: 0.1450435221195221, grad: -0.018003715202212334
before 1, rank: 1, weight: 0.1450435221195221
before 1, rank: 3, weight: 0.1450435221195221
before 1, rank: 2, weight: 0.1450435221195221
after 1, rank: 0, weight: 0.1450435221195221, grad: -0.03955963999032974
after 1, rank: 3, weight: 0.1450435221195221, grad: -0.03072114661335945
before 2, rank: 0, weight: 0.1450435221195221
before 2, rank: 3, weight: 0.1450435221195221
after 1, rank: 1, weight: 0.1450435221195221, grad: -0.03775426745414734
before 2, rank: 1, weight: 0.1450435221195221
after 1, rank: 2, weight: 0.1450435221195221, grad: -0.03235533833503723
before 2, rank: 2, weight: 0.1450435221195221
after 2, rank: 0, weight: 0.1450435221195221, grad: -0.06408560276031494
after 2, rank: 3, weight: 0.1450435221195221, grad: -0.04222358390688896
after 2, rank: 1, weight: 0.1450435221195221, grad: -0.056242190301418304
last, rank: 0, weight: 0.20912912487983704, grad: -0.06408560276031494
last, rank: 3, weight: 0.18726710975170135, grad: -0.04222358390688896
last, rank: 1, weight: 0.201285719871521, grad: -0.056242190301418304
after 2, rank: 2, weight: 0.1450435221195221, grad: -0.04413666948676109
last, rank: 2, weight: 0.1891801953315735, grad: -0.04413666948676109
```
Weights and grads do not seem to be synchronized.

## Expected behavior

Grads should be synchronized.

## Additional context

This seems to be a gap in DDP where it doesn’t support running backward twice? I couldn’t find any tests that use retain_graph=True with DDP.

The problem seems to be this line: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L530, which skips any gradient reduction. This variable is set to false after the first backward is done: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L1152 and then never set to True again since prepare_for_backward is not called anymore.

**Comment by rohan-varma** [comment]
Would we want to support running backwards specifically twice or are there more general cases where `.backward()` can be called `n` times?

**Comment by pritamdamania87** [comment]
> Would we want to support running backwards specifically twice or are there more general cases where .backward() can be called n times?

I think we should support running it as many times on par with what is supported for local training.

**Comment by rohan-varma** [comment]
@pritamdamania87, what do you think about a `DDP.backward(model, [loss])` API or the like to support this and calling backwards when not all parameters get grad in backwards pass? 

In this case, the API can support calling backwards twice by simply calling `self.reducer.prepare_for_backwards()` in between successive bwd calls. In the case for params that don't get gradient, we can traverse the autograd graph from the loss function and pre-mark those parameters as ready for reduction.

**Comment by pritamdamania87** [comment]
@rohan-varma Ideally we should try to support this without adding a new API. There are a few things `prepare_for_backward` does, one is to set a bunch of state (buckets ready, expect autograd hooks etc.) and another is unused parameter detection.

I'm wondering if this can be done in a way such that `finalize_backward()` resets all the state such that we can run another backward without calling `prepare_for_backward`. Regarding unused parameter detection, we can reuse the same state of `unused_parameters_` as computed by the last `prepare_for_backward`?

**Comment by rohan-varma** [comment]
> I'm wondering if this can be done in a way such that `finalize_backward()` resets all the state such that we can run another backward without calling `prepare_for_backward`. Regarding unused parameter detection, we can reuse the same state of `unused_parameters_` as computed by the last `prepare_for_backward`?

I think that this could work as well for calling backwards multiple times, since the same set of unused parameters should hold in later calls to backwards after a single forward pass. Although that probably wouldn't address support for calling backwards when not all outputs are used to compute loss, but that's a separate issue.

**Comment by rohan-varma** [comment]
> Regarding unused parameter detection, we can reuse the same state of `unused_parameters_` as computed by the last `prepare_for_backward`?

Actually realized that this may not work for some cases. What if you have:
```
a, b = forward()
b.sum().backward(retain_graph=true) # parameters used to compute `a` won't get gradient
(a + b).sum().backward() # now all params will get gradient
```
(here unused param is due to an output not being used in loss).

If we use the same set of `unused_parameters` then this would be incorrect because `a` would be unused, and then used in the second backwards.

I've tried to address this with the pass through function proposal in https://github.com/pytorch/pytorch/issues/55876 and moving `prepare_for_backward` to this passthrough function in https://github.com/pytorch/pytorch/pull/57081, would be grea to get some thoughts on the general approach.

**Comment by WhoisZihan** [comment]
@rohan-varma Hi Rohan, has your solution been accepted? It would be great to support multiple backward, is there any workaround for it now?
